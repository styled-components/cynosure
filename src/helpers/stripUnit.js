// @flow

/**
 * Strip the unit from a given CSS value, returning just the number. (or the original value if an invalid string was passed). Passing an optional boolean also returns the stripped unit.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   '--dimension': stripUnit('100px')
 *   '--dimension': `${stripUnit('100px`) - 10}${stripUnit('100px', true)[1]}`
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   --dimension: ${stripUnit('100px')}
 *   --dimension: ${stripUnit('100px`) - 10}${stripUnit('100px', true)[1]}
 * `
 *
 * // CSS in JS Output
 *
 * element {
 *   '--dimension': 100
 *   '--dimension': 1000px
 * }
 */

function stripUnit(
  value: string | number,
  returnUnit?: boolean,
): number | Array<number | string | undefined> | string {
  const unitlessValue = parseFloat(value)
  if (returnUnit) {
    const cssRegex = /^([+-]?(?:\d+|\d*\.\d+))([a-z]*|%)$/
    if (typeof value === 'number' || !value.match(cssRegex)) return [value, '']
    const unit = value.match(cssRegex)[2] ? value.match(cssRegex)[2] : undefined
    return [unitlessValue, unit]
  } else {
    if (isNaN(unitlessValue)) return value
    return unitlessValue
  }
}

export default stripUnit
