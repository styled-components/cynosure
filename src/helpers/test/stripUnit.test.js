// @flow
import stripUnit from '../stripUnit'

describe('stripUnit', () => {
  // Number Returns
  it('should strip px from whole values', () => {
    expect({ '--dimension': stripUnit('1px') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1px', true) }).toMatchSnapshot()
  })

  it('should strip px from values', () => {
    expect({ '--dimension': stripUnit('1.5px') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.5px', true) }).toMatchSnapshot()
  })

  it('should strip pt from whole values', () => {
    expect({ '--dimension': stripUnit('1pt') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1pt', true) }).toMatchSnapshot()
  })

  it('should strip pt from values', () => {
    expect({ '--dimension': stripUnit('1.5pt') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.5pt', true) }).toMatchSnapshot()
  })

  it('should strip pc from whole values', () => {
    expect({ '--dimension': stripUnit('1pc') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1pc', true) }).toMatchSnapshot()
  })

  it('should strip pc from values', () => {
    expect({ '--dimension': stripUnit('1.5pc') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.5pc', true) }).toMatchSnapshot()
  })

  it('should strip mm from whole values', () => {
    expect({ '--dimension': stripUnit('1mm') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1mm', true) }).toMatchSnapshot()
  })

  it('should strip mm from values', () => {
    expect({ '--dimension': stripUnit('1.5mm') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.5mm', true) }).toMatchSnapshot()
  })

  it('should strip q from whole values', () => {
    expect({ '--dimension': stripUnit('1q') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1q', true) }).toMatchSnapshot()
  })

  it('should strip q from values', () => {
    expect({ '--dimension': stripUnit('1.5q') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.5q', true) }).toMatchSnapshot()
  })

  it('should strip cm from whole values', () => {
    expect({ '--dimension': stripUnit('1cm') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1cm', true) }).toMatchSnapshot()
  })

  it('should strip cm from values', () => {
    expect({ '--dimension': stripUnit('1.5cm') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.5cm', true) }).toMatchSnapshot()
  })

  it('should strip in from whole values', () => {
    expect({ '--dimension': stripUnit('1in') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1in', true) }).toMatchSnapshot()
  })

  it('should strip in from values', () => {
    expect({ '--dimension': stripUnit('1.5in') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.5in', true) }).toMatchSnapshot()
  })

  it('should strip em from whole value', () => {
    expect({ '--dimension': stripUnit('1em') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1em', true) }).toMatchSnapshot()
  })

  it('should strip em from decimal values', () => {
    expect({ '--dimension': stripUnit('1.2em') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.2em', true) }).toMatchSnapshot()
  })

  it('should strip rem from whole values', () => {
    expect({ '--dimension': stripUnit('1rem') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1rem', true) }).toMatchSnapshot()
  })

  it('should strip rem from decimal values', () => {
    expect({ '--dimension': stripUnit('1.2rem') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.2rem', true) }).toMatchSnapshot()
  })

  it('should strip ex from whole values', () => {
    expect({ '--dimension': stripUnit('1ex') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1ex', true) }).toMatchSnapshot()
  })

  it('should strip ex from decimal values', () => {
    expect({ '--dimension': stripUnit('1.2ex') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.2ex', true) }).toMatchSnapshot()
  })

  it('should strip ch from whole values', () => {
    expect({ '--dimension': stripUnit('1ch') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1ch', true) }).toMatchSnapshot()
  })

  it('should strip ch from decimal values', () => {
    expect({ '--dimension': stripUnit('1.2ch') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('1.2ch', true) }).toMatchSnapshot()
  })

  it('should strip vh from whole values', () => {
    expect({ '--dimension': stripUnit('100vh') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('100vh', true) }).toMatchSnapshot()
  })

  it('should strip vh from decimal values', () => {
    expect({ '--dimension': stripUnit('33.33vh') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('33.33vh', true) }).toMatchSnapshot()
  })

  it('should strip vw from whole values', () => {
    expect({ '--dimension': stripUnit('100vw') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('100vw', true) }).toMatchSnapshot()
  })

  it('should strip vw from decimal values', () => {
    expect({ '--dimension': stripUnit('33.33vw') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('33.33vw', true) }).toMatchSnapshot()
  })

  it('should strip vmin from whole values', () => {
    expect({ '--dimension': stripUnit('100vmin') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('100vmin', true) }).toMatchSnapshot()
  })

  it('should strip vmin from decimal values', () => {
    expect({ '--dimension': stripUnit('33.33vmin') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('33.33vmin', true) }).toMatchSnapshot()
  })

  it('should strip vmax from whole values', () => {
    expect({ '--dimension': stripUnit('100vmax') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('100vmax', true) }).toMatchSnapshot()
  })

  it('should strip vmax from decimal values', () => {
    expect({ '--dimension': stripUnit('33.33vmax') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('33.33vmax', true) }).toMatchSnapshot()
  })

  it('should strip % from whole values', () => {
    expect({ '--dimension': stripUnit('80%') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('80%', true) }).toMatchSnapshot()
  })

  it('should strip % from decimal values', () => {
    expect({ '--dimension': stripUnit('33.3%') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('33.3%', true) }).toMatchSnapshot()
  })

  it('should return a unitless value when passed', () => {
    expect({ '--dimension': stripUnit('33') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('33', true) }).toMatchSnapshot()
    // $FlowFixMe
    expect({ '--dimension': stripUnit(33) }).toMatchSnapshot()
    // $FlowFixMe
    expect({ '--dimension': stripUnit(33, true) }).toMatchSnapshot()
  })

  it('should return invalid value when passed', () => {
    expect({ '--dimension': stripUnit('blah') }).toMatchSnapshot()
    expect({ '--dimension': stripUnit('blah', true) }).toMatchSnapshot()
  })
})
