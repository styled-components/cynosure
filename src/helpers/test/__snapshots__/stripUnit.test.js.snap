// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripUnit should return a unitless value when passed 1`] = `
Object {
  "--dimension": 33,
}
`;

exports[`stripUnit should return a unitless value when passed 2`] = `
Object {
  "--dimension": Array [
    33,
    "",
  ],
}
`;

exports[`stripUnit should return a unitless value when passed 3`] = `
Object {
  "--dimension": 33,
}
`;

exports[`stripUnit should return a unitless value when passed 4`] = `
Object {
  "--dimension": Array [
    33,
    undefined,
  ],
}
`;

exports[`stripUnit should return invalid value when passed 1`] = `
Object {
  "--dimension": "blah",
}
`;

exports[`stripUnit should return invalid value when passed 2`] = `
Object {
  "--dimension": Array [
    "blah",
    undefined,
  ],
}
`;

exports[`stripUnit should strip % from decimal values 1`] = `
Object {
  "--dimension": 33.3,
}
`;

exports[`stripUnit should strip % from decimal values 2`] = `
Object {
  "--dimension": Array [
    33.3,
    "%",
  ],
}
`;

exports[`stripUnit should strip % from whole values 1`] = `
Object {
  "--dimension": 80,
}
`;

exports[`stripUnit should strip % from whole values 2`] = `
Object {
  "--dimension": Array [
    80,
    "%",
  ],
}
`;

exports[`stripUnit should strip ch from decimal values 1`] = `
Object {
  "--dimension": 1.2,
}
`;

exports[`stripUnit should strip ch from decimal values 2`] = `
Object {
  "--dimension": Array [
    1.2,
    "ch",
  ],
}
`;

exports[`stripUnit should strip ch from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip ch from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "ch",
  ],
}
`;

exports[`stripUnit should strip cm from values 1`] = `
Object {
  "--dimension": 1.5,
}
`;

exports[`stripUnit should strip cm from values 2`] = `
Object {
  "--dimension": Array [
    1.5,
    "cm",
  ],
}
`;

exports[`stripUnit should strip cm from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip cm from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "cm",
  ],
}
`;

exports[`stripUnit should strip em from decimal values 1`] = `
Object {
  "--dimension": 1.2,
}
`;

exports[`stripUnit should strip em from decimal values 2`] = `
Object {
  "--dimension": Array [
    1.2,
    "em",
  ],
}
`;

exports[`stripUnit should strip em from whole value 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip em from whole value 2`] = `
Object {
  "--dimension": Array [
    1,
    "em",
  ],
}
`;

exports[`stripUnit should strip ex from decimal values 1`] = `
Object {
  "--dimension": 1.2,
}
`;

exports[`stripUnit should strip ex from decimal values 2`] = `
Object {
  "--dimension": Array [
    1.2,
    "ex",
  ],
}
`;

exports[`stripUnit should strip ex from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip ex from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "ex",
  ],
}
`;

exports[`stripUnit should strip in from values 1`] = `
Object {
  "--dimension": 1.5,
}
`;

exports[`stripUnit should strip in from values 2`] = `
Object {
  "--dimension": Array [
    1.5,
    "in",
  ],
}
`;

exports[`stripUnit should strip in from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip in from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "in",
  ],
}
`;

exports[`stripUnit should strip mm from values 1`] = `
Object {
  "--dimension": 1.5,
}
`;

exports[`stripUnit should strip mm from values 2`] = `
Object {
  "--dimension": Array [
    1.5,
    "mm",
  ],
}
`;

exports[`stripUnit should strip mm from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip mm from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "mm",
  ],
}
`;

exports[`stripUnit should strip pc from values 1`] = `
Object {
  "--dimension": 1.5,
}
`;

exports[`stripUnit should strip pc from values 2`] = `
Object {
  "--dimension": Array [
    1.5,
    "pc",
  ],
}
`;

exports[`stripUnit should strip pc from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip pc from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "pc",
  ],
}
`;

exports[`stripUnit should strip pt from values 1`] = `
Object {
  "--dimension": 1.5,
}
`;

exports[`stripUnit should strip pt from values 2`] = `
Object {
  "--dimension": Array [
    1.5,
    "pt",
  ],
}
`;

exports[`stripUnit should strip pt from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip pt from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "pt",
  ],
}
`;

exports[`stripUnit should strip px from values 1`] = `
Object {
  "--dimension": 1.5,
}
`;

exports[`stripUnit should strip px from values 2`] = `
Object {
  "--dimension": Array [
    1.5,
    "px",
  ],
}
`;

exports[`stripUnit should strip px from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip px from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "px",
  ],
}
`;

exports[`stripUnit should strip q from values 1`] = `
Object {
  "--dimension": 1.5,
}
`;

exports[`stripUnit should strip q from values 2`] = `
Object {
  "--dimension": Array [
    1.5,
    "q",
  ],
}
`;

exports[`stripUnit should strip q from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip q from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "q",
  ],
}
`;

exports[`stripUnit should strip rem from decimal values 1`] = `
Object {
  "--dimension": 1.2,
}
`;

exports[`stripUnit should strip rem from decimal values 2`] = `
Object {
  "--dimension": Array [
    1.2,
    "rem",
  ],
}
`;

exports[`stripUnit should strip rem from whole values 1`] = `
Object {
  "--dimension": 1,
}
`;

exports[`stripUnit should strip rem from whole values 2`] = `
Object {
  "--dimension": Array [
    1,
    "rem",
  ],
}
`;

exports[`stripUnit should strip vh from decimal values 1`] = `
Object {
  "--dimension": 33.33,
}
`;

exports[`stripUnit should strip vh from decimal values 2`] = `
Object {
  "--dimension": Array [
    33.33,
    "vh",
  ],
}
`;

exports[`stripUnit should strip vh from whole values 1`] = `
Object {
  "--dimension": 100,
}
`;

exports[`stripUnit should strip vh from whole values 2`] = `
Object {
  "--dimension": Array [
    100,
    "vh",
  ],
}
`;

exports[`stripUnit should strip vmax from decimal values 1`] = `
Object {
  "--dimension": 33.33,
}
`;

exports[`stripUnit should strip vmax from decimal values 2`] = `
Object {
  "--dimension": Array [
    33.33,
    "vmax",
  ],
}
`;

exports[`stripUnit should strip vmax from whole values 1`] = `
Object {
  "--dimension": 100,
}
`;

exports[`stripUnit should strip vmax from whole values 2`] = `
Object {
  "--dimension": Array [
    100,
    "vmax",
  ],
}
`;

exports[`stripUnit should strip vmin from decimal values 1`] = `
Object {
  "--dimension": 33.33,
}
`;

exports[`stripUnit should strip vmin from decimal values 2`] = `
Object {
  "--dimension": Array [
    33.33,
    "vmin",
  ],
}
`;

exports[`stripUnit should strip vmin from whole values 1`] = `
Object {
  "--dimension": 100,
}
`;

exports[`stripUnit should strip vmin from whole values 2`] = `
Object {
  "--dimension": Array [
    100,
    "vmin",
  ],
}
`;

exports[`stripUnit should strip vw from decimal values 1`] = `
Object {
  "--dimension": 33.33,
}
`;

exports[`stripUnit should strip vw from decimal values 2`] = `
Object {
  "--dimension": Array [
    33.33,
    "vw",
  ],
}
`;

exports[`stripUnit should strip vw from whole values 1`] = `
Object {
  "--dimension": 100,
}
`;

exports[`stripUnit should strip vw from whole values 2`] = `
Object {
  "--dimension": Array [
    100,
    "vw",
  ],
}
`;
